packages/core/src/tracing/instrumentation/cloud/aws-sdk/v2/sqs.js

```diff
@@ -355,53 +356,103 @@ function instrumentedReceiveMessage(ctx, originalReceiveMessage, originalArgs) {
+  if (
+    typeof messageBody === 'string' &&
+    messageBody.startsWith('{') &&
+    messageBody.includes('"Type":"Notification"') &&
+    /"X_INSTANA_/i.test(messageBody)
+  ) {
+    try {
+      const parsedBody = JSON.parse(messageBody);
+      if (parsedBody && parsedBody.MessageAttributes) {
+        tracingAttributes.traceId = readMessageAttributeWithFallback(
+          parsedBody.MessageAttributes,
+          sqsAttributeNames.TRACE_ID,
+          sqsAttributeNames.LEGACY_TRACE_ID
+        );
+        tracingAttributes.parentId = readMessageAttributeWithFallback(
+          parsedBody.MessageAttributes,
+          sqsAttributeNames.SPAN_ID,
+          sqsAttributeNames.LEGACY_SPAN_ID
+        );
+        tracingAttributes.level = readMessageAttributeWithFallback(
+          parsedBody.MessageAttributes,
+          sqsAttributeNames.LEVEL,
+          sqsAttributeNames.LEGACY_LEVEL
+        );
+      }
+    } catch (e) {
+      // The attempt to parse the message body as JSON failed, so this is not an SQS message resulting from an SNS
+      // notification (SNS-to-SQS subscription), in which case we are not interested in the body. Ignore the error and
+      // move on.
     }
   }
+  return tracingAttributes;
+}
 
-  return attributes;
+function readMessageAttributeWithFallback(attributes, key1, key2) {
+  const attribute =
+    tracingUtil.readAttribCaseInsensitive(attributes, key1) || tracingUtil.readAttribCaseInsensitive(attributes, key2);
+  if (attribute) {
+    // attribute.stringValue is used by SQS message attributes, attribute.Value is used by SNS-to-SQS.
+    return attribute.StringValue || attribute.Value;
+  }
+}
+
+/**
+ * Checks whether the given tracingAttributes object has at least one attribute set, that is, if there have been Instana
+ * message attributes present when converting message attributes into this object.
+ */
+function hasTracingAttributes(tracingAttributes) {
+  return tracingAttributes.traceId != null || tracingAttributes.parentId != null || tracingAttributes.level != null;
 }
 
 /**
  * Add extra info to the entry span after messages are received
  * @param {*} span The SQS Span
  * @param {*} data The data returned by the SQS API
- * @param {*} attributes The SQS Message Attributes
+ * @param {*} tracingAttributes The message attributes relevant for tracing
  */
-function configureEntrySpan(span, data, attributes) {
+function configureEntrySpan(span, data, tracingAttributes) {
   span.data.sqs.size = data.Messages.length;
   span.ts = Date.now();
 
-  const spanT = readAttribCaseInsensitive(attributes, sqsAttributeNames.TRACE_ID, sqsAttributeNames.LEGACY_TRACE_ID);
-  const spanP = readAttribCaseInsensitive(attributes, sqsAttributeNames.SPAN_ID, sqsAttributeNames.LEGACY_SPAN_ID);
-
-  if (spanT) {
-    span.t = spanT;
-  }
-
-  if (spanP) {
-    span.p = spanP;
+  if (tracingAttributes.traceId && tracingAttributes.parentId) {
```

Comment:
This actually changes the behavior slightly. Before, if only one of traceId or parentId was present, we would use that single attribute. Actually, a trace ID without a parent ID or vice versa is invalid and should be ignored IMO, so now this check only succeeds if both traceID and parentID are there. In practice, this should not make any difference, because any well behaved upstream tracer will either send both IDs or no ID at all.